ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -181,17 +170,93 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
 static int
 S3FileSystemTerminate(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session)
 {
-    S3_FILE_SYSTEM *fs;
+    S3_FILE_SYSTEM *fs = (S3_FILE_SYSTEM *)fileSystem;
 
     UNUSED(session); /* unused */
 
-    fs = (S3_FILE_SYSTEM *)fileSystem;
     delete (fs->connection);
     free(fs);
 
     return (0);
 }
 
+/*
+ * S3ObjectList --
+ *     Return a list of object names for the given location.
+ */
+static int
+S3ObjectList(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *bucket,
+  const char *prefix, char ***objectList, uint32_t *count)
+{
+    S3_FILE_SYSTEM *fs = (S3_FILE_SYSTEM *)fileSystem;
+    std::vector<std::string> objects;
+    int ret;
+    if (ret = fs->connection->ListObjects(std::string(bucket), std::string(prefix), objects) != 0)
+        return (ret);
+    *count = objects.size();
+
+    S3ObjectListAdd(fs->storage, objectList, objects, *count);
+
+    return (ret);
+}
+
+/*
+ * S3ObjectListSingle --
+ *     Return a single object name for the given location.
+ */
+static int
+S3ObjectListSingle(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *bucket,
+  const char *prefix, char ***objectList, uint32_t *count)
+{
+    S3_FILE_SYSTEM *fs = (S3_FILE_SYSTEM *)fileSystem;
+    std::vector<std::string> objects;
+    int ret;
+    if (ret = fs->connection->ListObjects(
+                std::string(bucket), std::string(prefix), objects, 1, true) != 0)
+        return (ret);
+    *count = objects.size();
+
+    S3ObjectListAdd(fs->storage, objectList, objects, *count);
+
+    return (ret);
+}
+
+/*
+ * S3ObjectListFree --
+ *     Free memory allocated by s3_directory_list.
+ */
+static int
+S3ObjectListFree(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, char **objectList, uint32_t count)
+{
+    (void)fileSystem;
+    (void)session;
+
+    if (objectList != NULL) {
+        while (count > 0)
+            free(objectList[--count]);
+        free(objectList);
+    }
+
+    return (0);
+}
+
+/*
+ * S3ObjectListAdd --
+ *     Add an entry to the directory list, growing as needed.
```

Comment:
This confuses me a little bit. Are we adding an entry to an existing structure or are we allocating memory for a structure?
Do we want to append the new `entries` to an existing `objectList`? Or do we want to allocate enough memory so `objectList` can contain the `objects`?

Reply:
Thanks, have updated the comment here! Let me know if it's clearer now.

Reply:
@fionaochee, what is the commit where you updated the comment?

Reply:
Commit https://github.com/wiredtiger/wiredtiger/pull/7487/commits/c9aaa3d9a56482cf84db5f221a51f00e48ba73ad.